app/src/components/templateEditor/language/languageUtils.ts

```diff
@@ -0,0 +1,86 @@
+import { LRParser } from "@lezer/lr";
+
+/**
+ * Extracts all variables from a templated string.
+ *
+ * @param parser - The parser for the templating language.
+ *  The parser should be a language parser that emits Variable nodes.
+ * @param text - The text to extract variables from.
+ *
+ * @returns An array of variable names.
+ */
+export const extractVariables = (parser: LRParser, text: string) => {
+  const tree = parser.parse(text);
+  const variables: string[] = [];
+  const cur = tree.cursor();
+  do {
+    if (cur.name === "Variable") {
+      variables.push(text.slice(cur.node.from, cur.node.to));
+    }
+  } while (cur.next());
+  return variables;
+};
+
+/**
+ * Formats a templated string with the given variables.
+ *
+ * The parser should be a language parser that emits Variable nodes as children of some parent node.
+ */
+export const format = ({
+  parser,
+  text,
+  variables,
+  postFormat,
+}: {
+  /**
+   * The parser for the templating language.
+   *
+   * Should be MustacheLikeTemplatingLanguage or FStringTemplatingLanguage.
```

Comment:
is there anyway to restrict this type? typeof MustacheLIke... etc.

Reply:
Not really, the output parser type is too complex, and technically this function works with any LRParser.

It just won't find variables correctly unless you use one of the two langs we created or any other lang that outputs Variable nodes